src/internal_telemetry/allocations/mod.rs

```diff
@@ -1,9 +1,53 @@
 //! Allocation tracking exposed via internal telemetry.
 
 mod allocator;
+use std::{
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Mutex,
+    },
+    thread,
+    time::Duration,
+};
+
+use arr_macro::arr;
+use metrics::gauge;
+use rand_distr::num_traits::ToPrimitive;
 
 use self::allocator::Tracer;
-pub(crate) use self::allocator::{AllocationGroupId, AllocationLayer, GroupedTraceableAllocator};
+
+pub(crate) use self::allocator::{
+    without_allocation_tracing, AllocationGroupId, AllocationLayer, GroupedTraceableAllocator,
+};
+
+use crossbeam_utils::CachePadded;
+
+const NUM_GROUPS: usize = 256;
+/// These arrays represent the allocations and deallocations for each group.
+/// We pad each Atomic to reduce false sharing effects.
+static GROUP_MEM_ALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
+    arr![CachePadded::new(AtomicU64::new(0)); 256];
+static GROUP_MEM_DEALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
+    arr![CachePadded::new(AtomicU64::new(0)); 256];
```

Comment:
AFAIU, each of these groups would be accessed by the same thread, right? If so, is there an advantage to this separation over grouping them something like this?
```rust
struct Stats {
    allocs: AtomicU64,
    deallocs: AtomicU64,
}

static GROUP_MEM_STATS: [CachePadded<Stats>; NumGroups];
```

Reply:
Will test the performance impact of SoA vs AoS 👍

Reply:
It's worth testing, but one thing I'll say is that this current design optimizes to avoid as much cross-thread contention as it reasonably can.

Given that (generally) allocations will happen as a result of a particular source generating events, and then be deallocated when those events land at a sink, we would usually expect two threads to be interacting with a given allocation group ID.

As is, this ensures that those threads are _not_ contending on the same atomic -- since atomic contention is a huge factor in the performance overhead of this -- but moving them together would bring that contention back to the forefront.